lib/ansible/modules/network/fortios/fortios_facts.py

```diff
@@ -0,0 +1,291 @@
+  returned: always
+  type: dict
+
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.connection import Connection
+from ansible.module_utils.network.fortimanager.common import FAIL_SOCKET_MSG
+from ansible.module_utils.six import iteritems
+
+
+class Factbase(object):
+    def __init__(self, module, fos, uri=None):
+        self.module = module
+        self.fos = fos
+        self.uri = uri
+        self.facts = dict()
+
+
+FACT_SYSTEM_SUBSETS = frozenset([
+    'system_current-admins_select',
+    'system_firmware_select',
+    'system_fortimanager_status',
+    'system_ha-checksums_select',
+    'system_interface_select',
+    'system_status_select',
+    'system_time_select',
+])
+
+
+class System(Factbase):
+    def populate_facts(self):
+        fos = self.fos
+        vdom = self.module.params['vdom']
+        if self.uri.startswith(tuple(FACT_SYSTEM_SUBSETS)):
+            resp = fos.monitor('system', self.uri[len('system_'):].replace('_', '/'), vdom=vdom)
+            self.facts.update({self.uri: resp})
+
+
+FACT_SUBSETS = dict(
+    system=System,
+)
+
+VALID_SUBSETS = frozenset(FACT_SUBSETS.keys())
+
+
+def login(data, fos):
+    host = data['host']
+    username = data['username']
+    password = data['password']
+    ssl_verify = data['ssl_verify']
+
+    fos.debug('on')
+    if 'https' in data and not data['https']:
+        fos.https('off')
+    else:
+        fos.https('on')
+
+    fos.login(host, username, password, verify=ssl_verify)
+
+
+def main():
+    fields = {
+        "host": {"required": True, "type": "str"},
+        "username": {"required": True, "type": "str"},
+        "password": {"required": True, "type": "str", "no_log": True},
+        "vdom": {"required": False, "type": "str", "default": "root"},
+        "https": {"required": False, "type": "bool", "default": True},
+        "ssl_verify": {"required": False, "type": "bool", "default": False},
+        'gather_subset': {"required": False, "type": "list", "default": ['system_status_select']}
+    }
+
+    module = AnsibleModule(argument_spec=fields,
+                           supports_check_mode=False)
+
+    legacy_mode = 'host' in module.params and module.params['host'] is not None and \
+                  'username' in module.params and module.params['username'] is not None and \
+                  'password' in module.params and module.params['password'] is not None
+
+    # for now only support local connection mode
+    if not legacy_mode:
+        module.fail_json(**FAIL_SOCKET_MSG)
+    else:
+        try:
+            from fortiosapi import FortiOSAPI
+        except ImportError:
+            module.fail_json(msg="fortiosapi module is required")
+
+        fos = FortiOSAPI()
+
+        login(module.params, fos)
+        gather_subset = module.params['gather_subset']
+
+        runable_subsets = set()
+
+        for subset in gather_subset:
```

Comment:
The fact collection logic is added in https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/network/common/facts/facts.py and is re-used by {{ network_os }}_facts implementations. Please have a look at https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/network/nxos/nxos_facts.py.

Reply:
@fortinetps Did you get a chance to look at this?^